File: apps/marketplace/packages/shared/src/schemas.ts
import { z } from 'zod';

export const Role = z.enum(['CUSTOMER', 'ARTISAN', 'ADMIN']);
export type Role = z.infer<typeof Role>;

export const BookingStatus = z.enum([
  'REQUESTED', 'ACCEPTED', 'REJECTED',
  'EN_ROUTE', 'IN_PROGRESS', 'COMPLETED', 'CANCELLED',
]);
export type BookingStatus = z.infer<typeof BookingStatus>;

export const DocKind = z.enum(['ID_FRONT', 'ID_BACK', 'SELFIE', 'TRADE_LICENSE']);

const phone = z.string().regex(/^\+\d{6,15}$/, 'Phone must be E.164 format (e.g. +2348012345678).');

export const OtpRequest = z.object({ phone });
export const OtpVerify = z.object({ phone, code: z.string().length(6) });
export const RefreshTokenReq = z.object({ refreshToken: z.string().min(10) });

export const SetRoleReq = z.object({ role: z.enum(['CUSTOMER', 'ARTISAN']) });

export const UpdateMeReq = z.object({
  fullName: z.string().min(1).max(80).optional(),
  avatarUrl: z.string().url().optional(),
  email: z.string().email().optional(),
});

export const ArtisanOnboardingReq = z.object({
  bio: z.string().max(500).optional(),
  yearsExperience: z.number().int().min(0).max(70),
  baseRateCents: z.number().int().min(0),
  hourlyRateCents: z.number().int().min(0),
  categoryIds: z.array(z.string()).min(1).max(5),
});

export const AvailabilityReq = z.object({
  available: z.boolean(),
  lat: z.number().min(-90).max(90).optional(),
  lng: z.number().min(-180).max(180).optional(),
});

export const SearchArtisansReq = z.object({
  category: z.string().optional(),
  lat: z.coerce.number().min(-90).max(90),
  lng: z.coerce.number().min(-180).max(180),
  radiusKm: z.coerce.number().min(0.1).max(50).default(10),
  minRating: z.coerce.number().min(0).max(5).optional(),
  maxPriceCents: z.coerce.number().int().min(0).optional(),
  availableNow: z.coerce.boolean().optional(),
  sort: z.enum(['distance', 'rating', 'price']).default('distance'),
  limit: z.coerce.number().int().min(1).max(50).default(20),
  cursor: z.string().optional(),
});

export const CreateBookingReq = z.object({
  artisanId: z.string(),
  categoryId: z.string(),
  description: z.string().min(5).max(2000),
  addressText: z.string().min(3).max(300),
  addressLat: z.number().min(-90).max(90),
  addressLng: z.number().min(-180).max(180),
  scheduledFor: z.string().datetime().optional(),
  isEmergency: z.boolean().default(false),
});

export const RejectBookingReq = z.object({
  reason: z.string().min(1).max(300),
});

export const CancelBookingReq = z.object({
  reason: z.string().min(1).max(300),
});

export const ReviewReq = z.object({
  rating: z.number().int().min(1).max(5),
  comment: z.string().max(1000).optional(),
});

export const SendMessageReq = z.object({
  body: z.string().min(1).max(2000),
  kind: z.enum(['TEXT', 'IMAGE', 'LOCATION']).default('TEXT'),
  s3Key: z.string().optional(),
  lat: z.number().min(-90).max(90).optional(),
  lng: z.number().min(-180).max(180).optional(),
});

export const RegisterPushTokenReq = z.object({
  token: z.string().min(10),
  platform: z.enum(['ios', 'android', 'web']),
});

export const ReportReq = z.object({
  reportedId: z.string(),
  bookingId: z.string().optional(),
  reason: z.string().min(3).max(120),
  details: z.string().max(2000).optional(),
});

export type ArtisanSearchResult = {
  id: string;
  fullName: string | null;
  avatarUrl: string | null;
  bio: string | null;
  avgRating: number;
  ratingCount: number;
  completedJobs: number;
  baseRateCents: number;
  hourlyRateCents: number;
  available: boolean;
  verified: boolean;
  distanceKm: number;
  currentLat: number | null;
  currentLng: number | null;
  categories: { id: string; slug: string; name: string }[];
};

File: apps/marketplace/apps/api-py/app/schemas.py
"""Pydantic request / response models — parity with @artisan/shared zod schemas."""
from __future__ import annotations

import re
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

PHONE_RE = re.compile(r"^\+\d{6,15}$")


class Base(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


# ---------- Auth ----------
class OtpRequest(Base):
    phone: str

    @field_validator("phone")
    @classmethod
    def _validate(cls, v: str) -> str:
        if not PHONE_RE.match(v):
            raise ValueError("Phone must be E.164 format, e.g. +2348012345678")
        return v


class OtpVerify(Base):
    phone: str
    code: str = Field(min_length=6, max_length=6)


class RefreshTokenReq(Base):
    refreshToken: str


class TokensResp(Base):
    accessToken: str
    refreshToken: str


class UserResp(Base):
    id: str
    phone: str
    role: Literal["CUSTOMER", "ARTISAN", "ADMIN"]
    fullName: str | None = None
    avatarUrl: str | None = None


class VerifyOtpResp(TokensResp):
    user: UserResp
    isNew: bool


# ---------- Profile ----------
class SetRoleReq(Base):
    role: Literal["CUSTOMER", "ARTISAN"]


class UpdateMeReq(Base):
    fullName: str | None = Field(default=None, max_length=80)
    avatarUrl: str | None = None
    email: str | None = None


class ArtisanOnboardingReq(Base):
    bio: str | None = Field(default=None, max_length=500)
    yearsExperience: int = Field(ge=0, le=70)
    baseRateCents: int = Field(ge=0)
    hourlyRateCents: int = Field(ge=0)
    categoryIds: list[str] = Field(min_length=1, max_length=5)


class AvailabilityReq(Base):
    available: bool
    lat: float | None = None
    lng: float | None = None


# ---------- Discovery ----------
class CategoryResp(Base):
    id: str
    slug: str
    name: str
    iconKey: str | None = None
    active: bool


class ArtisanSearchResp(Base):
    id: str
    fullName: str | None
    avatarUrl: str | None
    bio: str | None
    avgRating: float
    ratingCount: int
    completedJobs: int
    baseRateCents: int
    hourlyRateCents: int
    available: bool
    verified: bool
    distanceKm: float
    currentLat: float | None
    currentLng: float | None
    categories: list[CategoryResp] = Field(default_factory=list)


class ArtisanListResp(Base):
    results: list[ArtisanSearchResp]


# ---------- Booking ----------
class CreateBookingReq(Base):
    artisanId: str
    categoryId: str
    description: str = Field(min_length=5, max_length=2000)
    addressText: str = Field(min_length=3, max_length=300)
    addressLat: float
    addressLng: float
    scheduledFor: datetime | None = None
    isEmergency: bool = False


class RejectReq(Base):
    reason: str = Field(min_length=1, max_length=300)


class CancelReq(Base):
    reason: str = Field(min_length=1, max_length=300)


class ReviewReq(Base):
    rating: int = Field(ge=1, le=5)
    comment: str | None = Field(default=None, max_length=1000)


class BookingResp(Base):
    id: str
    status: str
    description: str
    addressText: str
    addressLat: float
    addressLng: float
    priceCents: int | None
    isEmergency: bool
    customerId: str
    artisanId: str
    categoryId: str
    createdAt: datetime


# ---------- Chat ----------
class SendMessageReq(Base):
    body: str = Field(min_length=1, max_length=2000)
    kind: Literal["TEXT", "IMAGE", "LOCATION"] = "TEXT"
    s3Key: str | None = None
    lat: float | None = None
    lng: float | None = None


class MessageResp(Base):
    id: str
    threadId: str
    senderId: str
    body: str
    kind: str
    createdAt: datetime


# ---------- Push ----------
class RegisterPushTokenReq(Base):
    token: str = Field(min_length=10)
    platform: Literal["ios", "android", "web"]
